### ISSUE TITLE
[no-misused-promises] checksVoidReturn=true doesn't catch the error when variadic arguments are used (like `...handlers: Array<() => void>`)

### ISSUE THREAD

**Issue by dko-slapdash** [typescript-eslint/typescript-eslint#4015]
This issue affects e.g. express used with new route handlers which are required to be non-async in their TS typings. I think in case of variadic arguments, the plugin verifies just the 1st argument and, if it's okay, doesn't even check the rest of the arguments.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
     "@typescript-eslint/no-misused-promises": "error",
  }
}
```

```TS
function get2(_h1: () => void, _h2: () => void) {}

function get(..._handlers: Array<() => void>) {}

get2(
  () => {},
  async () => {} // <-- fails, which is what we want
);

get(
  () => {},
  async () => {}  // <-- this must fail too, but it does not
);
```

**Expected Result**

Both get() and get2() calls must raise an eslint error.

**Actual Result**

Only get2() is treated as incorrect, and get() is fine - although it should fail too:

![image](https://user-images.githubusercontent.com/46383452/137565736-b5e3b343-c793-4e16-8c4b-63df065966b1.png)

**Additional Info**

**Versions**

| package                                   | version |
| ----------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.0.0` |
| `@typescript-eslint/parser`               | `5.0.0` |
| `TypeScript`                              | `4.4.4` |
| `ESLint`                                  | `8.0.1` |
| `node`                                    | `16.11.1` |
